Adds watchdog and live updater

MrsTonedOne opened the pull request:
This is a toolset to manage a production server of tgstation13 (or its forks). It includes an update script that is able to update the server without having to stop or shutdown the server (the update will take effect next round) and a script to start the server and restart it if it crashes (optional, requires registry tweaks to disable the Windows crash dialog system wide)

This will force a live tracking of the configured git repo, so no local modifications are allowed. (this is to avoid issues with merge conflicts and because I'm too lazy to make the script detect them)

This isn't live on the server yet, and I'm not sure if this 'counts' for the feature freeze since it doesn't relate to the game code at all.

But I figured I'd add this in for commenting while I'm at work and when I get back from work I'll apply it to the server.

~~Once I make it a bit more generic and fix certain edge case issues, It would be ready for inclusion. Its configurable and repo/fork agnostic.~~

Ready

MrsTonedOne committed:
Adds a production server toolkit

Chocobro commented:
I think this can be exempt from feature freeze since it's system maintenance

Miauw62 commented:
sort of sucks that this is windows-only but oh well.

MrStonedOne commented:
I made it as light on logic to make it easy as fuck to port to linux

Iamgoofball commented on tools/tgstation-server/config.bat:
```diff
@@ -0,0 +1,28 @@
+set REPO_URL=https://github.com/tgstation/-tg-station.git
+
+rem what map file to use. This should be the name of the dm, not dmm (and without the .dm part)
+set MAPFILE=metastation
```
this needs to be box by default

box is our default "guarenteed to work" map

MrStonedOne replied:
Oh, forgot to change it back from when I was testing that the build script
was overwriting the ticked map file correctly
On May 8, 2015 2:51 PM, "Iamgoofball" <notifications@github.com> wrote:

> In tools/tgstation-server/config.bat
> <https://github.com/tgstation/-tg-station/pull/9411#discussion_r29979876>:
>
> > @@ -0,0 +1,28 @@
> > +rem Server Tools configuration file. Lines starting with rem are comments and ignored.
> > +
> > +rem This must be set to the name of your dme without the .dme part. (should be fine leaving this alone unless you renamed the code)
> > +set PROJECTNAME=tgstation
> > +
> > +rem location of the repo.
> > +set REPO_URL=https://github.com/tgstation/-tg-station.git
> > +
> > +rem what map file to use. This should be the name of the dm, not dmm (and without the .dm part)
> > +set MAPFILE=metastation
>
> this needs to be box by default
>
> box is our default "guarenteed to work" map
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tgstation/-tg-station/pull/9411/files#r29979876>.
>

Iamgoofball replied:
sounds good

vista- commented:
If I have the time on the weekend, I'll make a *nix port.

d3athrow commented:
interesting implementation, dont know why you didnt just use nexis' watchdog though.

MrStonedOne commented:
@d3athrow mainly because I hate python as a language, mainly because this will just work in windows 7 without having to install a language, configure it, path it, then install like 50 million dependencies, and also because I want to be able to customize it.

This installs updates LIVE(kinda), no need to stop or restart dd, users experience no interruption, at the next round, the new update will be used during a normal world restart.

MrsTonedOne committed:
Minor tweak of the server tool. (and some more documentation)

vista- commented:
I doubt installing Python on the server would be a challenging task. Portability > convenience.

MrStonedOne commented:
To me? no To some guy who wants to "set up his own server" maybe.

Installing python isn't actually that hard, getting it in the path and installing a program's dependencies on the other hand....

Its just another item on a long list of items that anybody who looks at setting up a server sees and starts to get overwhelmed by.

Mine requires git for windows be installed (and byond of course).

After that, its click, wait, done.

optimumtact commented:
Watchdog isn't a required step however, it's an optional think as an ease of life feature for admins.

Either way, as long as we have something, we barely support linux as is with no bygex linux lib

MrsTonedOne committed:
Final tweaks to server tools + bot starter/watch dog

I mainly replaced some pauses with timeouts, and added a title to all the windows.

vista- commented:
@optimumtact I actually have a libbygex.so compiled, if you are interested in testing it, PM me on IRC (I might be offline as it is just a bouncer).

MrsTonedOne committed:
Improves server tools.
You can now specify a branch name to follow.
Server tools can now automatically push compiled changelogs to remote
Server tools will no longer fail to update properly if updated twice within one round.
Server tools will properly clean the git repo before updating.
Server tools will now properly copy over git's HEAD log so the game can detect its current revision.
Removed some debug lines that get left over and improved the output of the update script.

optimumtact commented:
This isn't getting merged because you besmirched the name of python
